### PR TITLE
CRUD APIs for report and report definition

### PR DESCRIPTION
**Pull request by zhongnansu**
*Issue #, if available:*
#20 #45 #46 
*Description of changes:*
- Added schema for report/report definition
- report API:
  - GET `api/reporting/reports?size=5&sortField=time_created&sortDirection=desc`
  - GET `api/reporting/reports/{reportId}`
  - DELETE `api/reporting/reports/{reportId}`

- report definition API:
  - POST `api/reporting/reportDefinition`
  - GET `api/reporting/reportDefinitions?sortField=time_created&sortDirection=desc&size=5`
  - GET `api/reporting/reportDefinitions/{reportDefinitionId}`
  - PUT `api/reporting/reportDefinitions/{reportDefinitionId}`
  - DELETE `api/reporting/reportDefinitions/{reportDefinitionId}`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: server/model/index.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import { schema } from '@kbn/config-schema';
+
+// TODO: needs update when we integrate csv feature
+const data_report_params = schema.object({
```
Comment (davidcui-amzn) [inline server/model/index.ts:19]: np: constants should be in all caps

Comment (davidcui-amzn) [inline server/model/index.ts:19]: how come these consts are snake case and the others in the file are camel case?

File: server/model/index.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  has_attachment: schema.boolean({ defaultValue: true }),
+  recipients: schema.arrayOf(schema.string(), { minSize: 1 }),
+});
+
+export const reportSchema = schema.object({
+  report_name: schema.string(),
+  report_source: schema.oneOf([
+    schema.literal('Dashboard'),
+    schema.literal('Visualization'),
+    schema.literal('Saved Search'),
```
Comment (davidcui-amzn) [inline server/model/index.ts:75]: From the UX this entry is listed as `Saved search` vs. `Saved Search`, the search is not capitalized

File: server/model/index.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  report_params: schema.conditional(
+    schema.siblingRef('report_source'),
+    'Saved Search',
+    data_report_params,
+    visual_report_params
+  ),
+
+  delivery: schema.maybe(
+    schema.object({
+      channel: schema.oneOf([
```
Comment (davidcui-amzn) [inline server/model/index.ts:92]: Since we only have Kibana and Email as options right now, I think it should be ok to leave out Slack and Chime

Comment (zhongnansu) [reply server/model/index.ts:92]: That's true, but it doesn't affect the schema definition here

File: server/model/index.ts (hunk 2 above)
Comment (zhongnansu) [reply server/model/index.ts:75]: will change

File: server/model/index.ts (hunk 1 above)
Comment (zhongnansu) [reply server/model/index.ts]: thx for pointing that out. Changed the `consts` to use camel case

File: server/routes/report.ts [hunk 4]
```diff
@@ -214,9 +208,7 @@ export default function (router: IRouter) {
           }
         );
         return response.ok({
-          body: {
-            data: esResp,
-          },
+          body: esResp._source,
```
Comment (davidcui-amzn) [inline server/routes/report.ts:211]: question: does this change the response returned from the API at all? Or is it refactoring for more elegant API work in back-end?

File: server/routes/report.ts [hunk 5]
```diff
@@ -230,16 +222,41 @@ export default function (router: IRouter) {
...
-  if (error.response) {
-    try {
-      const esErrorResponse = JSON.parse(error.response);
-      return esErrorResponse.reason || error.response;
-    } catch (parsingError) {
-      return error.response;
+  // Delete single report by id
+  router.delete(
+    {
+      path: `${API_PREFIX}/reports/{reportId}`,
```
Comment (davidcui-amzn) [inline server/routes/report.ts:229]: Think this route should be something like `${API_PREFIX}/delete_report_by_id/{reportId}`

File: server/routes/reportDefinition.ts [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+  REPORT_DEF_STATUS,
+  SCHEDULE_TYPE,
+  DELIVERY_CHANNEL,
+} from './utils/constants';
+
+export default function (router: IRouter) {
+  // Create report Definition
+  router.post(
+    {
+      path: `${API_PREFIX}/reportDefinition`,
```
Comment (davidcui-amzn) [inline server/routes/reportDefinition.ts:42]: what do you think about renaming this to something like `createReportDefinition`? It looks a bit more descriptive to me

File: server/routes/reportDefinition.ts [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+          body: parseEsErrorResponse(error),
+        });
+      }
+    }
+  );
+
+  // Update definition by id
+  router.put(
+    {
+      path: `${API_PREFIX}/reportDefinitions/{reportDefinitionId}`,
```
Comment (davidcui-amzn) [inline server/routes/reportDefinition.ts:106]: same comment here: maybe a more descriptive route name?

File: server/routes/reportDefinition.ts [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+          params
+        );
+
+        return response.ok({
+          body: esResp._id,
+        });
+      } catch (error) {
+        //@ts-ignore
+        context.reporting_plugin.logger.error(
+          `Fail to update report definition: ${error}`
```
Comment (davidcui-amzn) [inline server/routes/reportDefinition.ts:160]: np: Failed vs Fail

File: server/routes/reportDefinition.ts [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+          body: parseEsErrorResponse(error),
+        });
+      }
+    }
+  );
+
+  // get all report definitions details
+  router.get(
+    {
+      path: `${API_PREFIX}/reportDefinitions`,
```
Comment (davidcui-amzn) [inline server/routes/reportDefinition.ts:174]: same comment here about descriptive routes

File: server/routes/report.ts (hunk 4 above)
Comment (zhongnansu) [reply server/routes/report.ts:211]: yes, it extracts only the useful data for client side, gets rid of the redundant data fields returned by ES, such as `seq_num`, `version_num`, `id`

File: server/routes/reportDefinition.ts [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    }
+  );
+}
+
+function validateReportDefinition(reportDefinition: any) {
+  reportSchema.validate(reportDefinition);
+
+  const delivery = reportDefinition.delivery;
+  const trigger = reportDefinition.trigger;
+  const delivery_params = delivery.delivery_params;
```
Comment (davidcui-amzn) [inline server/routes/reportDefinition.ts:305]: np: camel case

File: server/routes/report.ts (hunk 5 above)
Comment (zhongnansu) [reply server/routes/report.ts:229]: https://restfulapi.net/resource-naming/
> URIs should not be used to indicate that a CRUD function is performed. URIs should be used to uniquely identify resources and not any action upon them. HTTP request methods should be used to indicate which CRUD function is performed.

File: server/routes/reportDefinition.ts (hunk 6 above)
Comment (zhongnansu) [reply server/routes/reportDefinition.ts:42]: uri naming shouldn't be descriptive of actions.

Comment (davidcui-amzn) [reply server/routes/reportDefinition.ts:42]: Gotcha, TIL

File: server/routes/report.ts (hunk 5 above)
Comment (davidcui-amzn) [reply server/routes/report.ts:229]: TIL

File: server/routes/reportDefinition.ts (hunk 8 above)
Comment (zhongnansu) [reply server/routes/reportDefinition.ts:160]: fixed

File: server/routes/reportDefinition.ts (hunk 10 above)
Comment (zhongnansu) [reply server/routes/reportDefinition.ts:305]: fixed, thx

File: server/model/index.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  url: schema.uri(),
+  window_width: schema.number({ defaultValue: 1200 }),
+  window_height: schema.number({ defaultValue: 800 }),
+  report_format: schema.oneOf([schema.literal('pdf'), schema.literal('png')]),
+});
+
+export const scheduleSchema = schema.object({
+  schedule_type: schema.oneOf([
+    schema.literal('Now'),
+    schema.literal('Future Date'),
```
Comment (davidcui-amzn) [inline server/model/index.ts:35]: Sorry i missed these but "Date" in Future date and "Based" in Cron based are not capitalized either

File: server/plugin.ts [hunk 12]
```diff
@@ -60,6 +61,7 @@ export class OpendistroKibanaReportsPlugin
 
     // Register server side APIs
     registerReportRoutes(router);
+    registerReportDefinitionRoutes(router);
```
Comment (seraphjiang) [inline server/plugin.ts:64]: try merge into one registerRoutes(router)
so we don't need to touch plugin.ts file when we introduce new router.

File: server/routes/reportDefinition.ts [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+          sortField: schema.maybe(schema.string()),
+          sortDirection: schema.maybe(schema.string()),
+        }),
+      },
+    },
+    async (
+      context,
+      request,
+      response
+    ): Promise<IKibanaResponse<any | ResponseError>> => {
```
Comment (seraphjiang) [inline server/routes/reportDefinition.ts:187]: anonymous method  is not friendly, in callstack when exception.

File: server/plugin.ts (hunk 12 above)
Comment (zhongnansu) [reply server/plugin.ts]: Make sense, fixed

**Pull request merged**